README.md
# NodeMediaClient-Android
[![](https://jitpack.io/v/NodeMedia/NodeMediaClient-Android.svg)](https://jitpack.io/#NodeMedia/NodeMediaClient-Android)   
A simple, high-performance, low-latency live streaming SDK.

## Features
### Play
* RTMP/RTSP/HLS/HTTP/KMP/UDP protocols
* FLV/MP4/fMP4/MKV/MPEGTS demuxers
* H264/H265 video decoders
* AAC/OPUS/G711/SPEEX/NELLYMOSER audio decoders
* Hardware Acceleration
* Low latency
* Delay elimination
* Take screenshot while playing
* Take record while playing, support mp4/flv/ts/mkv format 
* Compatible with flv_extension_id and Enhanced-Rtmp standards

### Publish
* RTMP/RTSP/HLS/HTTP/KMP/UDP protocols
* FLV/MPEGTS muxers
* H264/H265 video encoders 
* AAC audio encoder
* Hardware Acceleration
* Arbitrary video resolution
* Multiple output
* Compatible with flv_extension_id and Enhanced-Rtmp standards

## Install
### 1. Add the JitPack repository to your build file
```
dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
    repositories {
        google()
        mavenCentral()
        maven { url 'https://jitpack.io' }
    }
}
```

### 2. Add the dependency
```
dependencies {
    implementation 'com.github.NodeMedia:NodeMediaClient-Android:3.2.1'
}
```

## Play Live Streaming

### 1. Add permission INTERNET
```
<uses-permission android:name="android.permission.INTERNET" />
```

### 2. Setting up the layout
```
<FrameLayout
    android:id="@+id/video_view"
    android:layout_width="match_parent"
    android:layout_height="match_parent">
</FrameLayout>
```

### 3. Play the stream
```
private NodePlayer np;
    
@Override
protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    supportRequestWindowFeature(Window.FEATURE_NO_TITLE);
    setContentView(R.layout.activity_playview);

    FrameLayout vv = findViewById(R.id.video_view);
    np = new NodePlayer(this,"");
    np.attachView(vv);
    np.start("rtmp://192.168.0.2/live/demo");
}
    
@Override
protected void onDestroy() {
    super.onDestroy();
    np.detachView();
    np.stop();
}
```
That's it. Very simple!


## Publish Live Streaming
### 1. Request more permissions
```
<uses-feature android:name="android.hardware.camera.any" />

<uses-permission android:name="android.permission.INTERNET" />
<uses-permission android:name="android.permission.CAMERA" />
<uses-permission android:name="android.permission.RECORD_AUDIO" />
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
```

### 2. Permission to apply
```
private static final String[] PERMISSIONS = new String[]{
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.CAMERA,
        Manifest.permission.RECORD_AUDIO};
private static final int REQUEST_PERMISSION_CODE = 0XFF00;

@Override
protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    requestPermission();
    ……………………
}

private void requestPermission() {
    ActivityCompat.requestPermissions(MainActivity.this, PERMISSIONS, REQUEST_PERMISSION_CODE);
}

```

### 3. Setting up the layout
```
<FrameLayout
    android:id="@+id/camera_view"
    android:layout_width="match_parent"
    android:layout_height="match_parent">
</FrameLayout>
```

### 4.Start Publish
```
private NodePublisher np;

@Override
protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    supportRequestWindowFeature(Window.FEATURE_NO_TITLE);
    setContentView(R.layout.activity_publish_view);
    FrameLayout fl = findViewById(R.id.camera_view);

    np = new NodePublisher(this, "");
    np.setAudioCodecParam(NodePublisher.NMC_CODEC_ID_AAC, NodePublisher.NMC_PROFILE_AUTO, 48000, 1, 64_000);
    np.setVideoOrientation(NodePublisher.VIDEO_ORIENTATION_PORTRAIT);
    np.setVideoCodecParam(NodePublisher.NMC_CODEC_ID_H264, NodePublisher.NMC_PROFILE_AUTO, 480, 854, 30, 1_000_000);
    np.attachView(fl);
    np.openCamera(true);
    Button publishBtn = findViewById(R.id.publish_btn);
    publishBtn.setOnClickListener((v) -> {
        np.start("rtmp://192.168.0.2/live/demo");
    });
}

@Override
protected void onDestroy() {
    super.onDestroy();
    np.detachView();
    np.closeCamera();
    np.stop();
}
```

## License
A commercial license is required.  
[https://www.nodemedia.cn/product/nodemediaclient-android/](https://www.nodemedia.cn/product/nodemediaclient-android/)

## Business & Technical service
* QQ: 281269007
* Email: service@nodemedia.cn
* Telegram: https://t.me/illuspas


ChangeLog.md
## 3.2.1 - 2023-12-12
推流API可以通过设置音量来控制静音或者增益

## 3.2.0 - 2023-12-06
全面兼容Enhanced-Rtmp标准推流和播放h265编码

## 3.1.16 - 2023-07-10
优化播放重连策略
修复播放录制部分情况未写入结束包

## 3.1.15 - 2023-07-05
优化实现播放时录像

## 3.1.14 - 2023-06-28
优化实现播放时录像

## 3.1.13 - 2023-06-26
实现播放时录像

## 3.1.12 - 2023-06-7
完善点播api

## 3.1.11 - 2023-05-16
* 增加推流多输出,可用于多平台推流和直播录像

## 3.1.10 - 2023-05-12
* 修复一处音频问题

## 3.1.9 - 2023-05-12
* 修复异常
* 修复stop未清屏

## 3.1.8 - 2023-04-28
* 增加RTSP传输协议设置
* 增加HTTP referer/UA 设置
* 增加RTMP swfUrl/pageUrl 设置

## 3.1.7 - 2023-04-27
* 支持双声道降噪


nodemediaclient/src/main/java/cn/nodemedia/NodePlayer.java
/**
 * ©2023 NodeMedia.cn
 * <p>
 * Copyright © 2015 - 2023 NodeMedia.cn All Rights Reserved.
 */

package cn.nodemedia;

import android.content.Context;
import android.graphics.SurfaceTexture;
import android.view.Gravity;
import android.view.Surface;
import android.view.TextureView;
import android.view.ViewGroup;
import android.widget.FrameLayout;


public class NodePlayer implements TextureView.SurfaceTextureListener {
    static {
        System.loadLibrary("NodeMediaClient");
    }

    public static final int LOG_LEVEL_ERROR = 0;
    public static final int LOG_LEVEL_INFO = 1;
    public static final int LOG_LEVEL_DEBUG = 2;

    public static final String RTSP_TRANSPORT_UDP = "udp";
    public static final String RTSP_TRANSPORT_TCP = "tcp";
    public static final String RTSP_TRANSPORT_UDP_MULTICAST = "udp_multicast";
    public static final String RTSP_TRANSPORT_HTTP = "http";

    private static final String TAG = "NodeMedia.java";

    private OnNodePlayerEventListener onNodePlayerEventListener = null;
    private TextureView tv = null;
    private Context ctx;
    private long id;

    private FrameLayout.LayoutParams LP = new FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT,
            Gravity.CENTER);

    /**
     * 创建NodePlayer
     *
     * @param context Android context
     * @param license 授权码
     */
    public NodePlayer(Context context, String license) {
        id = jniInit(context, license);
        ctx = context;
    }

    @Override
    protected void finalize() {
        jniFree();
    }


    /**
     * 附加到视图
     *
     * @param vg ViewGroup的子类
     */
    public void attachView(ViewGroup vg) {
        if (this.tv == null) {
            this.tv = new TextureView(ctx);
            this.tv.setLayoutParams(LP);
            this.tv.setSurfaceTextureListener(this);
            this.tv.setKeepScreenOn(true);
            vg.addView(this.tv);
        }
    }

    /**
     * 返回当前的TextureView
     *
     * @return 当前的TextureView
     */
    public TextureView getTextureView() {
        return this.tv;
    }

    /**
     * 从视图中移除
     */
    public void detachView() {
        if (this.tv != null) {
            this.tv.setKeepScreenOn(false);
            this.tv = null;
        }
    }

    /**
     * 设置事件回调
     * @param listener
     */
    public void setOnNodePlayerEventListener(OnNodePlayerEventListener listener) {
        this.onNodePlayerEventListener = listener;
    }

    private void onEvent(int event, String msg) {
//        Log.d(TAG, "on Event: " + event + " Message:" + msg);
        if (this.onNodePlayerEventListener != null) {
            this.onNodePlayerEventListener.onEventCallback(this, event, msg);
        }
    }

    private native long jniInit(Context context, String license);

    private native void jniFree();

    /**
     * 开始播放
     *
     * @param url 播放的url
     * @return
     */
    public native int start(String url);

    /**
     * 停止播放
     *
     * @return
     */
    public native int stop();

    /**
     * 暂停或恢复点播视频播放
     *
     * @param pause 是否暂停
     * @return
     */
    public native int pause(boolean pause);

    /**
     * 时移
     *
     * @param pts 时移点，单位毫秒
     * @return
     */
    public native int seek(long pts);

    /**
     * 视频截图
     *
     * @param filename 保存的文件名，jpeg格式
     * @return
     */
    public native int screenshot(String filename);

    /**
     * 视频是否是点播回放
     *
     * @return 是否点播
     */
    public native boolean isVod();

    /**
     * 视频是否暂停了
     *
     * @return 是否暂停
     */
    public native boolean isPause();

    /**
     * 获取点播视频时长
     *
     * @return 单位毫秒
     */
    public native long getDuration();

    /**
     * 获取点播视频当前播放点
     *
     * @return 单位毫秒
     */
    public native long getCurrentPosition();

    /**
     * 获取点播视频缓冲点
     *
     * @return 单位毫秒
     */
    public native long getBufferPosition();

    /**
     * 获取点播视频缓冲百分比
     * @return 百分比
     */
    public native int getBufferPercentage();

    /**
     * 获取播放器是否真正在播放
     * @return
     */
    public native boolean isPlaying();

    /**
     * 设置日志等级
     *
     * @param logLevel 等级
     */
    public native void setLogLevel(int logLevel);

    /**
     * 设置缓存时长
     *
     * @param bufferTime 单位毫秒
     */
    public native void setBufferTime(int bufferTime);

    /**
     * 设置缩放模式
     *
     * @param mode 模式
     */
    public native void setScaleMode(int mode);

    /**
     * 设置视频surface
     *
     * @param surface 视频surface
     */
    public native void setVideoSurface(Surface surface);

    public native void setRTMPPageUrl(String rtmpPageUrl);

    public native void setRTMPSwfUrl(String rtmpSwfUrl);

    /**
     * 设置RTSP的传输协议， 默认是UDP
     * @param rtspTransport
     */
    public native void setRTSPTransport(String rtspTransport);

    /**
     * 设置HTTP Referer
     * @param httpReferer
     */
    public native void setHTTPReferer(String httpReferer);

    /**
     * 设置HTTP User-Agent
     * @param httpUserAgent
     */
    public native void setHTTPUserAgent(String httpUserAgent);
    /**
     * 设置视频解密密码
     *
     * @param cryptoKey 16字节密码
     */
    public native void setCryptoKey(String cryptoKey);

    /**
     * 设置音量
     *
     * @param volume 0.0 --- 1.0
     */
    public native void setVolume(float volume);

    /**
     * 设置是否开启硬件加速
     * @param enable 开关
     */
    public native void setHWAccelEnable(boolean enable);

    /**
     * 视频surface大小已改变
     */
    public native void resizeVideoSurface();

    public native void rotateVideo(int rotate);

    @Override
    public void onSurfaceTextureAvailable(SurfaceTexture surfaceTexture, int width, int height) {
        setVideoSurface(new Surface(surfaceTexture));
//        Log.i(TAG, "onSurfaceTextureAvailable: " + width + "x" + height);
    }

    @Override
    public void onSurfaceTextureSizeChanged(SurfaceTexture surfaceTexture, int width, int height) {
//        Log.i(TAG, "onSurfaceTextureSizeChanged: " + width + "x" + height);
        resizeVideoSurface();
    }

    @Override
    public boolean onSurfaceTextureDestroyed(SurfaceTexture surfaceTexture) {
//        Log.i(TAG, "onSurfaceTextureDestroyed");
//        setVideoSurface(null);
        return false;
    }

    @Override
    public void onSurfaceTextureUpdated(SurfaceTexture surfaceTexture) {
//        Log.d(TAG, "onSurfaceTextureUpdated");
    }

    public interface OnNodePlayerEventListener {
        void onEventCallback(NodePlayer player, int event, String msg);
    }

}

nodemediaclient/src/main/java/cn/nodemedia/NodePublisher.java
/**
 * ©2023 NodeMedia
 * <p>
 * Copyright © 2015 - 2023 NodeMedia.All Rights Reserved.
 */

package cn.nodemedia;

import android.content.Context;
import android.graphics.SurfaceTexture;
import android.opengl.GLSurfaceView;
import android.util.Size;
import android.view.Gravity;
import android.view.Surface;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.FrameLayout;
import androidx.annotation.NonNull;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraInfo;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.LifecycleOwner;
import com.google.common.util.concurrent.ListenableFuture;
import java.util.concurrent.ExecutionException;
import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class NodePublisher {
    static {
        System.loadLibrary("NodeMediaClient");
    }

    public static final int LOG_LEVEL_ERROR = 0;
    public static final int LOG_LEVEL_INFO = 1;
    public static final int LOG_LEVEL_DEBUG = 2;

    public static final int NMC_CODEC_ID_H264 = 27;
    public static final int NMC_CODEC_ID_H265 = 173;
    public static final int NMC_CODEC_ID_AAC = 86018;

    public static final int NMC_PROFILE_AUTO = 0;
    public static final int NMC_PROFILE_H264_BASELINE = 66;
    public static final int NMC_PROFILE_H264_MAIN = 77;
    public static final int NMC_PROFILE_H264_HIGH = 100;
    public static final int NMC_PROFILE_H265_MAIN = 1;
    public static final int NMC_PROFILE_AAC_LC = 1;
    public static final int NMC_PROFILE_AAC_HE = 4;
    public static final int NMC_PROFILE_AAC_HE_V2 = 28;
    public static final int NMC_PROFILE_AAC_LD = 22;
    public static final int NMC_PROFILE_AAC_ELD = 38;

    public static final int VIDEO_RC_CRF = 0;
    public static final int VIDEO_RC_ABR = 1;
    public static final int VIDEO_RC_CBR = 2;
    public static final int VIDEO_RC_VBV = 3;

    public static final int VIDEO_ORIENTATION_PORTRAIT = 0;
    public static final int VIDEO_ORIENTATION_LANDSCAPE_RIGHT = 1;
    public static final int VIDEO_ORIENTATION_LANDSCAPE_LEFT = 3;

    public static final int EffectorTextureTypeT2D = 0;
    public static final int EffectorTextureTypeEOS = 1;


    private static final String TAG = "NodeMedia.java";
    private OnNodePublisherEventListener onNodePublisherEventListener;
    private OnNodePublisherEffectorListener onNodePublisherEffectorListener;
    private GLCameraView glpv;
    private Camera mCamera;
    private Context ctx;
    private long id;
    private int fpsCount;
    private long fpsTime;
    private boolean isOpenFrontCamera = false;
    private int videoOrientation = Surface.ROTATION_0;
    private int videoWidth = 720;
    private int videoHeight = 1280;
    private int cameraWidth = 0;
    private int cameraHeight = 0;
    private int surfaceWidth = 0;
    private int surfaceHeight = 0;

    private final FrameLayout.LayoutParams LP = new FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT,
            Gravity.CENTER);

    public NodePublisher(@NonNull Context context, @NonNull String license) {
        ctx = context;
        id = jniInit(context, license);
    }

    public void setOnNodePublisherEventListener(OnNodePublisherEventListener onNodePublisherEventListener) {
        this.onNodePublisherEventListener = onNodePublisherEventListener;
    }

    public void setOnNodePublisherEffectorListener(OnNodePublisherEffectorListener onNodePublisherEffectorListener) {
        this.onNodePublisherEffectorListener = onNodePublisherEffectorListener;
    }

    public void attachView(@NonNull ViewGroup vg) {
        if (this.glpv == null) {
            this.glpv = new GLCameraView(this.ctx);
            this.glpv.setLayoutParams(LP);
            this.glpv.setKeepScreenOn(true);
            vg.addView(this.glpv);
        }
    }

    public void detachView() {
        if (this.glpv != null) {
            this.glpv.setKeepScreenOn(false);
            this.glpv = null;
            closeCamera();
            GPUImageDestroy();
        }
    }

    public void setVideoOrientation(int orientation) {
        this.videoOrientation = orientation;
    }

    public void openCamera(boolean frontCamera) {
        this.isOpenFrontCamera = frontCamera;
        ListenableFuture<ProcessCameraProvider> cameraProviderFuture = ProcessCameraProvider.getInstance(ctx);
        cameraProviderFuture.addListener(() -> {
            try {
                ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                bindImageAnalysis(cameraProvider, this.isOpenFrontCamera);
            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        }, ContextCompat.getMainExecutor(this.ctx));
    }

    public void closeCamera() {
        ListenableFuture<ProcessCameraProvider> cameraProviderFuture = ProcessCameraProvider.getInstance(ctx);
        cameraProviderFuture.addListener(() -> {
            try {
                ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                cameraProvider.unbindAll();
            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        }, ContextCompat.getMainExecutor(this.ctx));
    }

    public void switchCamera() {
        this.isOpenFrontCamera = !this.isOpenFrontCamera;
        closeCamera();
        openCamera(this.isOpenFrontCamera);
    }

    public Camera getCamera() {
        return mCamera;
    }

    public CameraInfo getCameraInfo() {
        return mCamera.getCameraInfo();
    }

    private void bindImageAnalysis(@NonNull ProcessCameraProvider cameraProvider, boolean front) {
        CameraSelector cameraSelector = front ? CameraSelector.DEFAULT_FRONT_CAMERA : CameraSelector.DEFAULT_BACK_CAMERA;

        Preview preview = new Preview.Builder()
                .setTargetResolution(new Size(videoWidth, videoHeight))
                .setTargetRotation(videoOrientation)
                .build();
        preview.setSurfaceProvider(this.glpv.getSurfaceProvider());
        mCamera = cameraProvider.bindToLifecycle((LifecycleOwner) this.ctx, cameraSelector  , preview);
    }

    private void onEvent(int event, String msg) {
//        Log.d(TAG, "on Event: " + event + " Message:" + msg);
        if (this.onNodePublisherEventListener != null) {
            this.onNodePublisherEventListener.onEventCallback(this, event, msg);
        }
    }

    private void onCreateEffector() {
        if (this.onNodePublisherEffectorListener != null) {
            this.onNodePublisherEffectorListener.onCreateEffector(this.ctx);
        }
    }

    private int onProcessEffector(int textureID) {
        if (this.onNodePublisherEffectorListener != null) {
            textureID = this.onNodePublisherEffectorListener.onProcessEffector(textureID, this.videoWidth, this.videoHeight);
        }
        return textureID;
    }

    private void onReleaseEffector() {
        if (this.onNodePublisherEffectorListener != null) {
            this.onNodePublisherEffectorListener.onReleaseEffector();
        }
    }

    protected void finalize() {
        jniFree();
    }

    private native long jniInit(Context context, String license);

    private native void jniFree();

    public native void setLogLevel(int logLevel);

    public native void setHWAccelEnable(boolean enable);

    public native void setDenoiseEnable(boolean enable);

    public native void setVideoFrontMirror(boolean mirror);

    public native void setCameraFrontMirror(boolean mirror);

    public native void setAudioCodecParam(int codec, int profile, int sampleRate, int channels, int bitrate);

    public native void setVideoCodecParam(int codec, int profile, int width, int height, int fps, int bitrate);

    public native void setVideoRateControl(int rc);

    public native void setKeyFrameInterval(int keyFrameInterval);

    /**
     * 设置视频解密密码
     *
     * @param cryptoKey 16字节密码
     */
    public native void setCryptoKey(@NonNull String cryptoKey);

    /**
     * 设置是否使用enhanced-rtmp 标准推流
     * @param enhancedRtmp
     */
    public native void setEnhancedRtmp(boolean enhancedRtmp);

    public native int addOutput(@NonNull String url);

    public native int removeOutputs();

    public native int start(@NonNull String url);

    public native int stop();

    public native void setEffectorTextureType(int type);

    private native int GPUImageCreate(int textureID);

    private native int GPUImageChange(int sw, int sh, int cw, int ch, int so, int co, boolean f);

    private native int GPUImageDraw(int textureID, float[] mtx, int len);

    private native int GPUImageDestroy();

    private native int GPUImageGenOESTextureID();

    private void onViewChange() {
        if (this.cameraWidth == 0 || this.cameraHeight == 0 || this.surfaceWidth == 0 || this.surfaceHeight == 0) {
            return;
        }
        WindowManager wm = (WindowManager) this.ctx.getSystemService(Context.WINDOW_SERVICE);
        int surfaceRotation = wm.getDefaultDisplay().getRotation();
        int sensorRotationDegrees = getCameraInfo().getSensorRotationDegrees(this.videoOrientation);
        GPUImageChange(this.surfaceWidth, this.surfaceHeight, this.cameraWidth, this.cameraHeight, surfaceRotation, sensorRotationDegrees, this.isOpenFrontCamera);
    }

    private class GLCameraView extends GLSurfaceView implements GLSurfaceView.Renderer {
        private static final String TAG = "NodeMedia.GLCameraView";

        private SurfaceTexture surfaceTexture;
        private int textureId = -1;
        private Context context;
        private float transformMatrix[] = new float[16];

        protected GLCameraView(Context context) {
            super(context);
            this.context = context;
            setEGLContextClientVersion(2);
            setRenderer(this);
            setRenderMode(GLSurfaceView.RENDERMODE_WHEN_DIRTY);
        }

        private Preview.SurfaceProvider getSurfaceProvider() {
            return request -> {
                Size resolution = request.getResolution();
                surfaceTexture.setDefaultBufferSize(resolution.getWidth(), resolution.getHeight());
                request.provideSurface(new Surface(surfaceTexture), ContextCompat.getMainExecutor(this.context), result -> {
                    result.getSurface().release();
                });
                this.queueEvent(() -> {
                    NodePublisher.this.cameraWidth = resolution.getWidth();
                    NodePublisher.this.cameraHeight = resolution.getHeight();
                    NodePublisher.this.onViewChange();
                });
            };
        }

        @Override
        public void onSurfaceCreated(GL10 gl10, EGLConfig eglConfig) {
            textureId = GPUImageGenOESTextureID();
            surfaceTexture = new SurfaceTexture(textureId);
            surfaceTexture.setOnFrameAvailableListener(surfaceTexture -> requestRender());
            NodePublisher.this.GPUImageCreate(textureId);
        }

        @Override
        public void onSurfaceChanged(GL10 gl10, int w, int h) {
            NodePublisher.this.surfaceWidth = w;
            NodePublisher.this.surfaceHeight = h;
            NodePublisher.this.onViewChange();
        }

        @Override
        public void onDrawFrame(GL10 gl10) {
            surfaceTexture.updateTexImage();
            surfaceTexture.getTransformMatrix(transformMatrix);
            NodePublisher.this.GPUImageDraw(textureId, transformMatrix, transformMatrix.length);
        }
    }

    public interface OnNodePublisherEffectorListener {

        void onCreateEffector(Context context);

        int onProcessEffector(int textureID, int width, int height);

        void onReleaseEffector();

    }

    public interface OnNodePublisherEventListener {
        void onEventCallback(NodePublisher publisher, int event, String msg);
    }
}